### PR TITLE
HLG: get_all_external_keys()

### PR DESCRIPTION
**Pull request by madsbk**
This PR introduces `HighLevelGraph.get_all_external_keys()`, which makes it possible to cull high level graphs that contains `ShuffleLayer` without materializing them.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

cc. @rjzamora

### CODE REVIEW

**Review by rjzamora — APPROVED** [review]
The approach here makes sense to me @madsbk . I was particularly focued on the shuffle changes as I reviewed, but the use of external dependencies for culling seems like a good idea to me - Thanks for working on this!

File: dask/dataframe/shuffle.py [hunk 1]
```diff
@@ -72,6 +72,9 @@ def __init__(
         self.meta_input = meta_input
         self.parts_out = parts_out or range(npartitions)
 
+    def get_output_keys(self):
```
[the remark this answers is not in the document]
Comment (madsbk) [reply dask/dataframe/shuffle.py:75]: Yes, implementing layer specific `get_output_keys()` will be essential to avoid materialization.

**Comment by madsbk** [comment]
@jrbourbeau I think this is ready to be merged

**Review by jrbourbeau — APPROVED** [review]
Thanks @madsbk!

**Pull request merged**